### ISSUE TITLE
Unable to print with Ricoh MFP with legacy usb://

### ISSUE THREAD

**Issue by bfg01** [OpenPrinting/cups#510]
Printer is Ricoh SP 377SFNwx. Using openSUSE Leap 15.4, installed from scratch (no upgrade).

Got the PPD from here:
https://openprinting.org/printer/Ricoh/Ricoh-SP_377SFNwX

Added the printer with "lpadmin"; it succeeded with no errors (exit code 0).
But when trying to print anything either with any GUI app or "lp" command, the print job stays hung infinitely, with no activity from printer nor any console output. I always have to use the "cancel" command to remove the print job.

Tried getting the logs with this process: "cupsctl LogLevel=debug2", try printing, check /var/log/cups/error_log.
I attached the log with .txt suffix here.

Could someone help please?
Thanks beforehand.

[error_log.txt](https://github.com/OpenPrinting/cups/files/9872117/error_log.txt)

**Comment by jsmeix** [comment]
@bfg01 
I did not check the details but on first glance your
https://github.com/OpenPrinting/cups/files/9872117/error_log.txt
contains (excerpts):
```
D [26/Oct/2022:11:14:06 -0500] [Job 19] Auto-typing file...
D [26/Oct/2022:11:14:06 -0500] [Job 19] Request file type is application/vnd.cups-pdf-banner.
```
Did you really print a file of type `application/vnd.cups-pdf-banner`?
Those are the files in /usr/share/cups/banners
when you specify banner pages for printing
and the cups-filters test page /usr/share/cups/data/testprint

I ask because you wrote
```
print anything either with any GUI app or "lp" command
```
which indicates you print e.g. a normal plain text file like
```
# lp -d ricoh1 /etc/os-release
```

**Comment by jsmeix** [comment]
https://www.ricoh-usa.com/en/products/pd/equipment/printers-and-copiers/multifunction-printers-copiers/sp-377sfnwx-black-and-white-laser-multifunction-printer
reads
```
SP 377SFNwX
Black and White Laser Multifunction Printer
...
Printer Specifications
...
Printer Languages        PCL6, PCL5e
```
so the Ghostscript driver 'pxlmono' for PCL6 should work for it
and also the Ghostscript driver 'ljet4' for PCL5e should work
where for the latter there is the generic PPD file
/usr/share/cups/model/OpenPrintingPPDs/ghostscript/Generic-PCL_5e_Printer.ljet4.ppd.gz

In contrast to 'pxlmono' the Ghostscript driver 'ljet4'
does not emit HP PJL commands but only plain PCL5e.
Perhaps the SP 377SFNwX might not like the PJL commands
('PJL' is not mentioned in the above specifications)
and then perhaps plain PCL5e may work?

**Comment by zdohnal** [comment]
Hi @bfg01 ,

I've checked the error_log you uploaded, I downloaded the PPD, installed a queue with file backend, tried to print and check the logs on my machine:

Your log regarding the job ends around the lines:

```
3861 d [26/Oct/2022:11:14:08 -0500] select_timeout: JobHistoryUpdate=0
3862 D [26/Oct/2022:11:14:08 -0500] [Job 19] @PJL SET RESOLUTION=600
3863 D [26/Oct/2022:11:14:08 -0500] [Job 19] @PJL SET BITSPERDOT=1
3864 D [26/Oct/2022:11:14:08 -0500] [Job 19] @PJL SET DUPLEX=ON
3865 D [26/Oct/2022:11:14:08 -0500] [Job 19] @PJL SET BINDING=LONGEDGE
3866 D [26/Oct/2022:11:14:08 -0500] [Job 19] <job data> \033%-12345X@PJL RESET
3867 d [26/Oct/2022:11:14:08 -0500] select_timeout: JobHistoryUpdate=0
3868 I [26/Oct/2022:11:14:09 -0500] Expiring subscriptions...
3869 d [26/Oct/2022:11:14:09 -0500] select_timeout: JobHistoryUpdate=0
3870 d [26/Oct/2022:11:14:09 -0500] select_timeout(0): 13 seconds to write dirty config/state files
3871 I [26/Oct/2022:11:14:22 -0500] Saving printers.conf...
...
```

There is about 10 seconds delay in the log and the next messages related to the job are about it being cancelled by you:

```
4077 D [26/Oct/2022:11:14:22 -0500] [Job 19] time-at-completed=1666800862
...
4084 I [26/Oct/2022:11:14:22 -0500] [Job 19] Job canceled by \"test1\"
...
4091 I [26/Oct/2022:11:14:22 -0500] [Job 19] Canceled by "test1".
```

If I run a job on my queue with Ricoh PPD, I get the following log (around the same lines you have):

```
Oct 27 09:19:15 fedora cupsd[1306]: [Job 7] @PJL SET RESOLUTION=600
Oct 27 09:19:15 fedora cupsd[1306]: [Job 7] @PJL SET BITSPERDOT=1
Oct 27 09:19:15 fedora cupsd[1306]: [Job 7] @PJL SET DUPLEX=ON
Oct 27 09:19:15 fedora cupsd[1306]: [Job 7] @PJL SET BINDING=LONGEDGE
Oct 27 09:19:15 fedora cupsd[1306]: [Job 7] <job data> \033%-12345X@PJL RESET
Oct 27 09:19:15 fedora cupsd[1306]: [Job 7] -----------
Oct 27 09:19:15 fedora cupsd[1306]: [Job 7] New page: %%Page: 2 2
Oct 27 09:19:15 fedora cupsd[1306]: [Job 7] Found: %%Page: 2 2
Oct 27 09:19:15 fedora cupsd[1306]: [Job 7] --> Output goes to the FIFO buffer now.
```

Renderer is started for you as a ghostscript command:

```
3851 D [26/Oct/2022:11:14:08 -0500] [Job 19] Starting renderer with command: \"(printf \'\\033%%-12345X@PJL\\n@PJL JOB\\n\');
3852 D [26/Oct/2022:11:14:08 -0500] [Job 19] (gs -q -dBATCH -dPARANOIDSAFER -dNOPAUSE -dNOMEDIAATTRS -dNOINTERPOLATE -sDEVICE=pxlmono  -r600x600 -sPAPERSIZE=letter -sOutputFile=- -  | sed \"s/\\(\\xf8.\\xc0\\)\\x00\\xf8\\x34/                  \\1\\x01\\xf8\\x35/g\" | sed \"s/^\\x1b\\x25-12345X//\" | sed \"s/\\xc1\\x01\\x00\\xf8\\x31\\x44/\\x44/g\");
3853 D [26/Oct/2022:11:14:08 -0500] [Job 19] (printf \'@PJL\\n@PJL EOJ\\n\\033%%-12345X\')\"
```

, so my guess is there is a hiccup in gs command execution on OpenSUSE, because renderer never finishes. The other difference is that your queue uses ```pdftops``` from poppler:

```
D [26/Oct/2022:11:14:06 -0500] [Job 19] Printer make and model: RICOH RICOH SP 377SFNwX PXL
D [26/Oct/2022:11:14:06 -0500] [Job 19] Running command line for pstops: pstops 19 test1 testprint 1 \' finishings=3 job-uuid=urn:uuid:5c7e5c25-1c94-3a76-7468-707df94bfb0a job-originating-host-name=localhost date-time-at-creation= date-ti
me-at-processing= time-at-creation=1666800846 time-at-processing=1666800846 document-name-supplied=testprint\'
D [26/Oct/2022:11:14:06 -0500] [Job 19] Using image rendering resolution 600 dpi
D [26/Oct/2022:11:14:06 -0500] [Job 19] Running command line for pdftops: pdftops -level1 -origpagesizes -nocenter -r 600 -optimizecolorspace /var/spool/cups/tmp/foomatic-rDFYfv -
D [26/Oct/2022:11:14:06 -0500] [Job 19] Started filter pdftops (PID 6346)
...
D [26/Oct/2022:11:14:07 -0500] [Job 19] %Produced by poppler pdftops version: 22.01.0 (http://poppler.freedesktop.org)
```

while mine runs ghostscript:

```
Oct 27 09:19:14 fedora cupsd[1306]: [Job 7] Printer make and model: RICOH RICOH SP 377SFNwX PXL
Oct 27 09:19:14 fedora cupsd[1306]: [Job 7] Running command line for pstops: pstops 7 anonymous \'Test Page\' 1 \' job-uuid=urn:uuid:0b774dc4-4be7-3b0f-470e-5d2697e3ad7b job-originating-host-name=localhost date-time-at-creation= date-tim>
Oct 27 09:19:14 fedora cupsd[1306]: [Job 7] Using image rendering resolution 600 dpi
Oct 27 09:19:14 fedora cupsd[1306]: [Job 7] Running command line for gs: gs -q -dNOPAUSE -dBATCH -dSAFER -dNOMEDIAATTRS -sstdout=%stderr -sDEVICE=ps2write -dShowAcroForm -sOUTPUTFILE=%stdout -sProcessColorModel=DeviceGray -sColorConversi>
Oct 27 09:19:14 fedora cupsd[1306]: [Job 7] Started filter gs (PID 30955)
...
Oct 27 09:19:15 fedora cupsd[1306]: [Job 7] %%Invocation: gs -q -dNOPAUSE -dBATCH -dSAFER -dNOMEDIAATTRS -sstdout=? -sDEVICE=ps2write -dShowAcroForm -sOUTPUTFILE=? -sProcessColorModel=DeviceGray -sColorConversionStrategy=Gray -dLanguageLevel=3 -r600 -dCompressFonts=false -dNoT3CCITT -dNOINTERPOLATE ?
```

To sum it up, there might be a hiccup in gs caused by something from poppler's pdftops.

Note: If the printer doesn't like specific PJL commands, I would expect renderer to finish and the error appears on the printer itself, or just by printing nothing, if all filters and backend finishes.

It would be great if you contacted your distro support - I've tried to reproduce the issue on the latest version and the rendering seems to work in comparison to your logs.

**Comment by zdohnal** [comment]
Note: I did print a testpage, since @jsmeix noticed the bannertopdf filter (I totally overlooked that :) )

**Comment by jsmeix** [comment]
FWIW:
For me the plain print job data processing
also works with all the above listed PPDs
but I do not have that printer so I cannot verify
how things behave with that printer at the USB.

My home-office system is not openSUSE Leap 15.4 from scratch
but something upgraded from older openSUSE Leap 15.x plus
some newer software like CUPS 2.4.2 (Leap 15 has CUPS 2.2.7).
Testing plain print job data processing
on openSUSE Leap 15.4 from scratch
would need some more time.

**Comment by jsmeix** [comment]
@bfg01 

PCL5e printers should be able to print plain ASCII text as is
but there are crippled PCL printers out there that won't
and I don't know how reliable the info at
https://openprinting.org/printer/Ricoh/Ricoh-SP_377SFNwX
actually is - i.e. whether or not
```
Printer supports PJL.
Printer supports direct text printing with the 'us-ascii' charset.
```
was actually verified.

So to test that try out if
```
echo -en "\rHello\r\f" | lp -d ricoh1 -o raw
```
makes the printer output a sheet of paper
with 'Hello' printed on it, cf.
https://en.opensuse.org/SDB:Installing_a_Printer#USB_printers

**Comment by bfg01** [comment]
I narrowed the problem.
First, sorry everyone, I should have done more tests...

I was actually to print raw, plain text, PDF, and even LibreOffice docs with legacy usb://... and the OpenPrinting PPD.

The issue: the CUPS testing page /usr/share/cups/data/testprint
This is seemingly the only one (until now) giving the problem.

I also tested this on Debian 11, also got the problem with the testing page, and attached the log here as txt as well.

Also tried @jsmeix 's suggestion of trying the two Generic-PCL_5e_Printer.ljet4{,d}.ppd.gz drivers on Leap. Surprisingly, these ones did **not** give the problem with CUPS testing page...

[error_log-Debian.txt](https://github.com/OpenPrinting/cups/files/9882704/error_log-Debian.txt)

**Comment by bfg01** [comment]
I narrowed the problem.
First, sorry everyone, I should have done more tests...

I was actually to print raw, plain text, PDF, and even LibreOffice docs with legacy usb://... and the OpenPrinting PPD.

The issue: the CUPS testing page /usr/share/cups/data/testprint
This is seemingly the only one (until now) giving the problem.

I also tested this on Debian 11, also got the problem with the testing page, and attached the log here as txt as well.

Also tried @jsmeix 's suggestion of trying the two Generic-PCL_5e_Printer.ljet4{,d}.ppd.gz drivers on Leap. Surprisingly, these ones did **not** give the problem with CUPS testing page...

[error_log-Debian.txt](https://github.com/OpenPrinting/cups/files/9882704/error_log-Debian.txt)

**Comment by zdohnal** [comment]
@bfg01 

> I was actually to print raw, plain text, PDF, and even LibreOffice docs with legacy usb://... and the OpenPrinting PPD.
> 

Thanks, good to know! So PJL is not completely broken, but the printer itself doesn't like the generated test page itself in PCL6 language + PJL commands.

> The issue: the CUPS testing page /usr/share/cups/data/testprint This is seemingly the only one (until now) giving the problem.

The file is actually a template, which is used to get a generated PDF via ```bannertopdf``` filter - the same filter is used for printing banners (the specific pages with words like 'Classified' or 'Secret' at the first page of the job or the last - f.e. if you print a documetn with 1 page and you add a starting banner, you print 2 pages...) - do banners work for you?

In Evince (the default PDF viewer in GNOME) you can find banner settings once you choose the printer in the dialog, go to the Job tab and set something in ```Add Cover Page``` scope.

If this does not work, then we can say ```bannertopdf``` generates output incompatible with the driver+printer.

> 
> I also tested this on Debian 11, also got the problem with the testing page, and attached the log here as txt as well.
> 

Based on the logs the renderer is ghostscript in Debian instead of Poppler's pdftops, so this crosses the renderer out. 

> Also tried @jsmeix 's suggestion of trying the two Generic-PCL_5e_Printer.ljet4{,d}.ppd.gz drivers on Leap. Surprisingly, these ones did **not** give the problem with CUPS testing page...

Neither of them uses PJL commands like PXL driver does... have you tried pxlmono PCL6 (PXL) generic driver which @jsmeix mentioned as well? 

In Fedora it lives under the driver uri ```foomatic:Generic-PCL_6_PCL_XL_Printer-pxlmono.ppd```, which you can use with ```-m``` option in ```lpadmin``` command:

```
$ lpadmin -p <name> -v <device_uri> -m foomatic:Generic-PCL_6_PCL_XL_Printer-pxlmono.ppd -E
```

You can list all drivers by ```lpinfo -m``` and check whether the driver uri is the same at OpenSUSE.

> 
> @zdohnal what about this still pending issue [OpenPrinting/ipp-usb#59](https://github.com/OpenPrinting/ipp-usb/issues/59) ? Thanks.

@alexpevzner is the author of the project and looks after it, he will look into it once he has time.

**Comment by bfg01** [comment]
Now I tried with the pxlmono PCL6 generic driver. Results: the very same as with the OpenPrinting driver (link left in OP).

The problem seems to be with the banners, as you suggested:
--tried with the CUPS testing page, printer makes an "error beep" and then print job either gets cancelled or marked as completed, with nothing actually being done
--tried with PDFs (the same ones I had already tested), but setting starting banner on Evince, printer did the same error beep but this time printed the document's pages

I had never heard about "banners" in printing until now; but if a "banner" is indeed an apart page, then printer is refusing to print them, and seemingly CUPS testing page is a whole banner page actually...

**Comment by zdohnal** [comment]
@bfg01 this sounds like an issue I had with older cups-filters, where bannertopdf was giving me trouble with testpage and banners, but in my case PDF document format was affected - the upgrade to the newer version helped. Here the issue is connected to PCL6 (PXL), since generic PCL 5e works for you. The model in OpenPrinting PPD database has even plain PCL 5 version of the driver (provided by Ricoh as well) with driver uri:

```
foomatic-db-ppds/Ricoh/PCL5/Ricoh-SP_377SFNwX_PCL5.ppd.gz
```

I know the device specification mentions only PCL6 and PCL 5e, but the driver was provided by Ricoh employee, so it might work as well.

In general regarding PCL6/PXL driver, it would be great if you tried the latest cups-filters and cups versions (Debian 11 does not have the latest - at least CUPS is 2.3.3op2, which is about one year old... the current one is 2.4.2 and cups-filters 1.28.16), whether you can hit the issue - and if you can, please ask on OpenPrinting/cups-filters for help.

**Comment by jsmeix** [comment]
Leap 15 and SLES 15 have cups-filters-1.25.0

The OBS Printing project has currently cups-filters-1.28.15
but that needs a newer libqpdf28 which is available
for Leap 15 and can be installed e.g. with
```
# zypper in libqpdf28
```
in addition to libqpdf26 that is normally used in Leap 15.

libqpdf26 contains
```
/usr/lib64/libqpdf.so.26
/usr/lib64/libqpdf.so.26.0.2
```
and libqpdf28 contains
```
/usr/lib64/libqpdf.so.28
/usr/lib64/libqpdf.so.28.3.1
```

When libqpdf28 is there, cups-filters-1.28.15 RPMs
from the OBS Printing project can be downloaded
and installed e.g. from
http://download.opensuse.org/repositories/Printing/
for various SUSE and openSUSE versions.
